rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,452 @@
+
+### Deleting data
+
+For this design, we are assuming that objects will never get deleted because the client will only be following transactions involving a small number of public keys. These transactions cannot be deleted because they are components used to prove the existence of transaction. We could delete blocks that failed to reach finality and the transaction involving them. However, the concept is outside of the scope of this RFC.
+
+### Adding Arbitrary Data
+
+Since a node can go offline, it can miss gossipped objects. If this happens, a client would be interested in manually adding these objects into the client storage system. This storage system has the flexibility to do this.
+
+For adding arbitrary transactions, the client could simply add it to the `transactions` database. We can also include the previous receipt chain hashes involving that transaction and use these components and add them to the `receipt_chain` database. The user can add arbitrary blocks along with their transactions into the `blocks_to_transactions` table, `block` and `transactions` database. When we add a block whose block length is greater than root length, the block will be added through the `transition_frontier_controller`. Namely, the block will be added to the `network_transition_writer` pipe in `Coda_networking.ml`.
+
+<a href="bootstrap"></a>
+
+### Running Bootstrap
+
+If a client is offline for long time and they have to bootstrap, then they would not have a `transition_frontier`. Therefore, all the transitions in the `transition_frontier` would have their `consensus_status` be considered to `UNKNOWN`. Also, the transactions would not be in an `transaction_pool` anymore, so all the transaction's in `User_command.mem` would be `false`.
+
+## Data Flow
+
+This section will discuss how data flows between the daemon and the databases.
+
+Below is a diagram of how the components in this RFC will depend on each other:
+
+![Client storage](../docs/res/client_storage.dot.png)
+
+### Archive Process
+
+To offload many database writes that has to be conducted by the daemon, we are going to delegate the writes to another process, which will be called an Archive Process. Whenever the `transition_frontier` and the `transaction_pool` makes an update, they will send the diff representing the update to the process. The diffs of the `transition_frontier` that the archive process will receive will be similar to `Transition_frontier` extension diffs. Currently, these diffs contain breadcrumbs and we cannot serialize breadcrumbs, so the diffs that clients will be receiving will be the following:
```

Comment:
How is the data written out in another process? Based on our discussion today we said it could use GraphQL subscriptions -- @nholland94 believes it should just make a POST request to the archive process directly. We can resolve this in person tomorrow.

Reply:
We should make it clear that we are not coupling the archive process to any OCamlisms directly and also make it clear that the API process and database are not coupled to specific choices

Reply:
I will make it clear. I will also emphasize that the API process should know what database it is reading from